zerver/lib/bulk_create.py

```diff
@@ -145,3 +157,124 @@ def bulk_create_streams(realm: Realm,
+    `prob_upvote`, provided there is another recipient of the message left to upvote.
+    Repeat the process for a different emoji with probability `prob_repeat`.
+
+    If the number of emojis or users is small, there is a chance the above process
+    will produce multiple reactions with the same user and emoji, so group the
+    reactions by emoji code and user profile and then return one reaction from
+    each group.
+    '''
+    for p in (prob_reaction, prob_repeat, prob_upvote):
+        if p > 1 or p < 0:
+            raise ValueError('Probability argument must be between 0 and 1.')
+    EMOJI_NAME = 0
+    EMOJI_CODE = 1
+    emojis = itertools.cycle(emojis)
+    if not users:
+        if message.recipient.type == Recipient.PERSONAL:
+            users = [
+                UserProfile.objects.get(recipient__id=message.recipient.id),
+                message.sender]
+        else:
+            users = [
+                subscription.user_profile
+                for subscription in Subscription.objects.filter(
+                    recipient=message.recipient)]
+        if not users:
+            return []
+    users = itertools.cycle(users)
+    reactions = []
+    if random.random() < prob_reaction:
+        emoji: Tuple[str, str] = next(emojis)
+        user: UserProfile = next(users)
+        reactions.append(Reaction(
+            user_profile=user,
+            message=message,
+            emoji_name=emoji[EMOJI_NAME],
+            emoji_code=emoji[EMOJI_CODE],
+            reaction_type=Reaction.UNICODE_EMOJI
+        ))
+        users_reacted = {user.id}
+        user = next(users)
+        while random.random() < prob_upvote and user.id not in users_reacted:
+            reactions.append(Reaction(
+                user_profile=user,
+                message=message,
+                emoji_name=emoji[EMOJI_NAME],
+                emoji_code=emoji[EMOJI_CODE],
+                reaction_type=Reaction.UNICODE_EMOJI
+            ))
+            users_reacted.add(user.id)
+            user = next(users)
+
+        while random.random() < prob_repeat:
+            emoji = next(emojis)
+            user = next(users)
+            reactions.append(Reaction(
+                user_profile=user,
+                message=message,
+                emoji_name=emoji[EMOJI_NAME],
+                emoji_code=emoji[EMOJI_CODE],
+                reaction_type=Reaction.UNICODE_EMOJI
+            ))
+            users_reacted = {user.id}
+            user = next(users)
+            while random.random() < prob_upvote and user.id not in users_reacted:
+                reactions.append(Reaction(
+                    user_profile=user,
+                    message=message,
+                    emoji_name=emoji[EMOJI_NAME],
+                    emoji_code=emoji[EMOJI_CODE],
+                    reaction_type=Reaction.UNICODE_EMOJI
+                ))
+                users_reacted.add(user.id)
+                user = next(users)
```

Comment:
This code can be deduplicated with the right loop structure.  I'll make the change.